File: lib/java/src/main/java/io/github/brenbar/uapi/internal/ValidateHeaders.java
package io.github.brenbar.uapi.internal;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static io.github.brenbar.uapi.internal.Prepend.prepend;

public class ValidateHeaders {
    static List<_ValidationFailure> validateHeaders(
            Map<String, Object> headers, Map<String, _UFieldDeclaration> parsedRequestHeaders, _UFn functionType) {
        final var validationFailures = new ArrayList<_ValidationFailure>();

        for (final var entry : headers.entrySet()) {
            final var header = entry.getKey();
            final var headerValue = entry.getValue();
            final var field = parsedRequestHeaders.get(header);
            if (field != null) {
                final var thisValidationFailures = field.typeDeclaration.validate(headerValue, null, functionType.name,
                        List.of());
                final var thisValidationFailuresPath = thisValidationFailures.stream()
                        .map(e -> new _ValidationFailure(prepend(header, e.path), e.reason, e.data)).toList();
                validationFailures.addAll(thisValidationFailuresPath);
            }
        }

        return validationFailures;
    }
}

File: lib/py/msgpact/internal/validation/ValidateHeaders.py
from typing import TYPE_CHECKING
from ...internal.validation.ValidationFailure import ValidationFailure

if TYPE_CHECKING:
    from ..types.VFieldDeclaration import VFieldDeclaration
    from ..types.VFn import VFn


def validate_headers(headers: dict[str, object], parsed_request_headers: dict[str, 'VFieldDeclaration'], function_type: 'VFn') -> list['ValidationFailure']:
    from ...internal.validation.ValidateContext import ValidateContext

    validation_failures = []

    for header, header_value in headers.items():
        field = parsed_request_headers.get(header)
        if field:
            this_validation_failures = field.type_declaration.validate(
                header_value, ValidateContext(None, function_type.name))
            this_validation_failures_path = [
                ValidationFailure([header] + e.path, e.reason, e.data)
                for e in this_validation_failures
            ]
            validation_failures.extend(this_validation_failures_path)

    return validation_failures
